src/NBitcoin/BlockStake.cs

```diff
@@ -270,16 +270,57 @@ public override Transaction CreateTransaction(byte[] bytes)
         }
     }
 
+    public interface ISmartContractBlockHeader
+    {
+        uint256 HashStateRoot { get; set; }
+
+        uint256 ReceiptRoot { get; set; }
+
+        Bloom LogsBloom { get; set; }
+    }
+
     /// <summary>
     /// A POS block header, this will create a work hash based on the X13 hash algos.
     /// </summary>
 #pragma warning disable 618
-    public class PosBlockHeader : BlockHeader
+    public class PosBlockHeader : BlockHeader, ISmartContractBlockHeader
 #pragma warning restore 618
     {
+        // Determines whether this object should serialize the new fields associated with smart contracts.
+        public bool HasSmartContractFields => (this.version & 0x10000000 /* ExtendedHeaderBit */) != 0;
+
         /// <inheritdoc />
         public override int CurrentVersion => 7;
 
+        private ushort extendedHeaderSize => (ushort)(hashStateRootSize + receiptRootSize + this.logsBloom.GetCompressedSize());
+
+        /// <summary>
+        /// Root of the state trie after execution of this block. 
+        /// </summary>
+        private uint256 hashStateRoot;
+        public uint256 HashStateRoot { get { return this.hashStateRoot; } set { this.hashStateRoot = value; } }
+        private static int hashStateRootSize = (new uint256()).GetSerializeSize();
```

Comment:
Question, does this represent a "proper" `uint256`s size? i.e, not one that is "empty"... ?

Reply:
`uint256` is always 32 bytes.